### ISSUE TITLE
[Bug]: Inference with the same inputs changes to different output values after many model invocations

### ISSUE THREAD

**Issue by Dobiasd** [openvinotoolkit/openvino#28824]
### OpenVINO Version

2024.6.0

### Operating System

Other (Please specify in description)

### Device used for inference

CPU

### Framework

None

### Model used

ResNet50

### Issue description

When running the same forward pass many times, the output suddenly changes to different values.

Here is a minimal (dockerized) example to reproduce the problem:

`main.py`

```python3
#!/usr/bin/env python3

import datetime
import logging
import threading
import time

import keras
import numpy as np
import openvino as ov
import requests
from flask import Flask
from flask_restful import Resource, Api
from openvino import CompiledModel


def load_model() -> CompiledModel:
    base = keras.applications.ResNet50(input_shape=(224, 224, 3), weights="imagenet", include_top=True)
    image_input = keras.layers.Input(shape=(224, 224, 3), name="input_layer")
    ov_model = ov.convert_model(keras.Model(inputs=image_input, outputs=base(image_input), name="image_model"))
    return ov.Core().compile_model(ov_model, "CPU")


np.random.seed(0)
image = np.random.rand(224, 224, 3)
model = load_model()
iteration = 0


class Controller(Resource):
    def get(self):
        preprocessed_images = [image.copy()]
        result = model(np.array(preprocessed_images))[0]
        global iteration
        iteration = iteration + 1
        print(f"{datetime.datetime.now()}: {iteration}, {result[0, 0]}", flush=True)


logging.getLogger("werkzeug").setLevel(logging.WARNING)
app = Flask("foo")
Api(app).add_resource(Controller, "/foo/")

threading.Thread(target=lambda: app.run(port=8080), daemon=True, name="foo").start()
time.sleep(1)
while True:
    requests.get("http://127.0.0.1:8080/foo", timeout=30)
```

Command:

```bash
docker build --rm --progress=plain .
```

Example output 1 (ran on `Intel(R) Xeon(R) Silver 4114 CPU @ 2.20GHz`):

```
[...]
#8 15.25 2025-02-04 17:30:19.986595: 1, 0.00015759124653413892
#8 15.28 2025-02-04 17:30:20.018549: 2, 0.00015759124653413892
#8 15.31 2025-02-04 17:30:20.047465: 3, 0.00015759124653413892
[...]
#8 465.6 2025-02-04 17:37:50.374849: 14762, 0.00015759124653413892
#8 465.7 2025-02-04 17:37:50.406194: 14763, 0.00015759124653413892
#8 465.7 2025-02-04 17:37:50.436870: 14764, 0.00015759124653413892
#8 466.0 2025-02-04 17:37:50.735640: 14765, 0.00015886615437921137
#8 466.0 2025-02-04 17:37:50.770898: 14766, 0.00015886615437921137
#8 466.1 2025-02-04 17:37:50.800891: 14767, 0.00015886615437921137
[...]
```

Example output 2 (ran on `Intel(R) Xeon(R) Silver 4114 CPU @ 2.20GHz`):

```
#8 15.12 2025-02-04 18:54:19.707492: 1, 0.00015759124653413892
#8 15.15 2025-02-04 18:54:19.738060: 2, 0.00015759124653413892
#8 15.18 2025-02-04 18:54:19.766186: 3, 0.00015759124653413892
[...]
#8 458.6 2025-02-04 19:01:43.206777: 14762, 0.00015759124653413892
#8 458.6 2025-02-04 19:01:43.236077: 14763, 0.00015759124653413892
#8 458.7 2025-02-04 19:01:43.265731: 14764, 0.00015759124653413892
#8 459.0 2025-02-04 19:01:43.566637: 14765, 0.0
#8 459.0 2025-02-04 19:01:43.600268: 14766, 0.0
#8 459.0 2025-02-04 19:01:43.629758: 14767, 0.0
[...]
```

Example output 3 (ran on `Intel(R) Core(TM) i7-14700K`)

```
[...]
#8 8.141 2025-02-04 17:43:08.254667: 1, 0.000157591188326478
#8 8.153 2025-02-04 17:43:08.266280: 2, 0.000157591188326478
#8 8.164 2025-02-04 17:43:08.277242: 3, 0.000157591188326478
[...]
#8 173.4 2025-02-04 17:45:53.514144: 14806, 0.000157591188326478
#8 173.4 2025-02-04 17:45:53.525131: 14807, 0.000157591188326478
#8 173.4 2025-02-04 17:45:53.536286: 14808, 0.000157591188326478
#8 173.5 2025-02-04 17:45:53.617902: 14809, 0.00015934312250465155
#8 173.5 2025-02-04 17:45:53.629846: 14810, 0.00016571232117712498
#8 173.5 2025-02-04 17:45:53.640727: 14811, 0.00016571232117712498
#8 173.5 2025-02-04 17:45:53.651473: 14812, 0.00016571232117712498
[...]
```

Example output 4 (ran on `Intel(R) Core(TM) i7-14700K`)

```
[...]
#8 8.591 2025-02-04 18:16:08.593288: 1, 0.000157591188326478
#8 8.601 2025-02-04 18:16:08.604064: 2, 0.000157591188326478
#8 8.612 2025-02-04 18:16:08.614221: 3, 0.000157591188326478
[...]
#8 169.6 2025-02-04 18:18:49.633181: 14854, 0.000157591188326478
#8 169.6 2025-02-04 18:18:49.643533: 14855, 0.000157591188326478
#8 169.7 2025-02-04 18:18:49.654100: 14856, 0.000157591188326478
#8 169.7 2025-02-04 18:18:49.728072: 14857, 1.0
#8 169.7 2025-02-04 18:18:49.740769: 14858, 1.0
#8 169.7 2025-02-04 18:18:49.751848: 14859, 1.0

[...]
```

This problem haunted us in PROD (more complex application, different model, and different data, of course), and it was quite some work to distill this minimal example. I hope it helps to find and fix the cause of the problem.

### Relevant log output

```shell

```

### Issue submission checklist

- [x] I'm reporting an issue. It's not a question.
- [x] I checked the problem with the documentation, FAQ, open issues, Stack Overflow, etc., and have not found a solution.
- [x] There is reproducer code and related data files such as images, videos, models, etc.

**Comment by Bheema-Shanker-Neyigapula** [comment]
.take

**Comment by rkazants** [comment]
Hi @Dobiasd,

I was unable to reproduce the issue on my local machine without docker.

Could you please try to specify additional parameter `share_weights=False` to `ov.convert_model`? It secures from uninitialized weights:

```
    ov_model = ov.convert_model(keras.Model(inputs=image_input, outputs=base(image_input), name="image_model"), share_weights=False)
```

Best regards,
Roman

**Comment by Bheema-Shanker-Neyigapula** [comment]
Hi @Dobiasd 

<html>
<body>
<html><head></head><body><p>Here’s a modified version of your script with fixes to mitigate the issue:</p>
<h3><strong>Fixes Applied</strong></h3>
<ol>
<li>
<p><strong>Forcing Single-Threaded Execution</strong></p>
<ul>
<li>Disabling OpenVINO multi-threading (<code inline="">INFERENCE_NUM_THREADS=1</code>).</li>
<li>Running Flask in non-threaded mode.</li>
</ul>
</li>
<li>
<p><strong>Handling Denormal Values</strong></p>
<ul>
<li>Enabling OpenVINO's <code inline="">DENORMALS_ZERO</code> mode to avoid floating-point instability.</li>
</ul>
</li>
<li>
<p><strong>Ensuring a Fresh Model Instance on Each Inference Call</strong></p>
<ul>
<li>Reloading the model before each inference prevents OpenVINO runtime drift.</li>
</ul>
</li>
<li>
<p><strong>Using Explicit NumPy Dtype (<code inline="">float32</code>)</strong></p>
<ul>
<li>Ensuring consistent input precision.</li>
</ul>
</li>
</ol>
<hr>
<h3><strong>Fixed Code (<code inline="">main.py</code>)</strong></h3>
<pre><code class="language-python">#!/usr/bin/env python3

import datetime
import logging
import time
import threading

import keras
import numpy as np
import openvino as ov
import requests
from flask import Flask
from flask_restful import Resource, Api
from openvino import CompiledModel

# Ensure deterministic behavior
np.random.seed(0)
image = np.random.rand(224, 224, 3).astype(np.float32)

# Enable OpenVINO debugging and denormal value handling
import os
os.environ["OV_LOG_LEVEL"] = "DEBUG"
os.environ["OPENVINO_DENORMALS_ZERO"] = "1"

def load_model() -&gt; CompiledModel:
    """Loads and compiles the ResNet50 model for OpenVINO"""
    base = keras.applications.ResNet50(input_shape=(224, 224, 3), weights="imagenet", include_top=True)
    image_input = keras.layers.Input(shape=(224, 224, 3), name="input_layer")
    
    # Convert to OpenVINO format
    ov_model = ov.convert_model(keras.Model(inputs=image_input, outputs=base(image_input), name="image_model"))

    # Create OpenVINO Core with single-threaded execution
    core = ov.Core()
    core.set_property("CPU", {"INFERENCE_NUM_THREADS": "1"})  # Force single-threading
    return core.compile_model(ov_model, "CPU", {"INFERENCE_PRECISION_HINT": "f32"})  # Use FP32 for stability

iteration = 0

class Controller(Resource):
    def get(self):
        """Handles inference requests, ensuring a fresh model instance each time"""
        global iteration
        model = load_model()  # Reload the model each request to avoid runtime corruption
        preprocessed_images = [image.copy()]
        result = model(np.array(preprocessed_images))[0]
        
        iteration += 1
        print(f"{datetime.datetime.now()}: {iteration}, {result[0, 0]}", flush=True)

# Suppress Flask logging
logging.getLogger("werkzeug").setLevel(logging.WARNING)

app = Flask("foo")
api = Api(app)
api.add_resource(Controller, "/foo/")

# Run Flask in single-threaded mode
threading.Thread(target=lambda: app.run(port=8080, threaded=False), daemon=True, name="foo").start()

# Wait for Flask to start
time.sleep(1)

# Continuous request loop
while True:
    requests.get("http://127.0.0.1:8080/foo", timeout=30)
</code></pre>
<hr>
<h3><strong>Summary of Fixes</strong></h3>

Issue | Fix
-- | --
Multi-threading instability | Forced single-threaded OpenVINO execution + Flask running with threaded=False
Floating-point drift / denormals | Enabled OPENVINO_DENORMALS_ZERO=1
Inference instability over time | Re-loading the model on each inference call
Precision variability | Explicitly using INFERENCE_PRECISION_HINT=f32


<p>Try running this version and let me know if the issue persists!</p></body></html>
</body>
</html>

**Comment by Bheema-Shanker-Neyigapula** [comment]
Hi @Dobiasd 

Here’s a modified version of your script with fixes to mitigate the issue:

### Fixes Applied

1. **Forcing Single-Threaded Execution**  
   - Disabled OpenVINO multi-threading by setting `INFERENCE_NUM_THREADS=1`.  
   - Flask is now running in non-threaded mode.

2. **Handling Denormal Values**  
   - Enabled OpenVINO's `DENORMALS_ZERO` mode to avoid floating-point instability.

3. **Ensuring a Fresh Model Instance on Each Inference Call**  
   - The model is now reloaded before each inference to prevent OpenVINO runtime drift.

4. **Using Explicit NumPy Dtype (`float32`)**  
   - Ensured consistent input precision.

---

### Fixed Code (`main.py`)
```python
#!/usr/bin/env python3

import datetime
import logging
import time
import threading

import keras
import numpy as np
import openvino as ov
import requests
from flask import Flask
from flask_restful import Resource, Api
from openvino import CompiledModel

# Ensure deterministic behavior
np.random.seed(0)
image = np.random.rand(224, 224, 3).astype(np.float32)

# Enable OpenVINO debugging and denormal value handling
import os
os.environ["OV_LOG_LEVEL"] = "DEBUG"
os.environ["OPENVINO_DENORMALS_ZERO"] = "1"

def load_model() -> CompiledModel:
    """Loads and compiles the ResNet50 model for OpenVINO"""
    base = keras.applications.ResNet50(input_shape=(224, 224, 3), weights="imagenet", include_top=True)
    image_input = keras.layers.Input(shape=(224, 224, 3), name="input_layer")
    
    # Convert to OpenVINO format
    ov_model = ov.convert_model(keras.Model(inputs=image_input, outputs=base(image_input), name="image_model"))

    # Create OpenVINO Core with single-threaded execution
    core = ov.Core()
    core.set_property("CPU", {"INFERENCE_NUM_THREADS": "1"})  # Force single-threading
    return core.compile_model(ov_model, "CPU", {"INFERENCE_PRECISION_HINT": "f32"})  # Use FP32 for stability

iteration = 0

class Controller(Resource):
    def get(self):
        """Handles inference requests, ensuring a fresh model instance each time"""
        global iteration
        model = load_model()  # Reload the model each request to avoid runtime corruption
        preprocessed_images = [image.copy()]
        result = model(np.array(preprocessed_images))[0]
        
        iteration += 1
        print(f"{datetime.datetime.now()}: {iteration}, {result[0, 0]}", flush=True)

# Suppress Flask logging
logging.getLogger("werkzeug").setLevel(logging.WARNING)

app = Flask("foo")
api = Api(app)
api.add_resource(Controller, "/foo/")

# Run Flask in single-threaded mode
threading.Thread(target=lambda: app.run(port=8080, threaded=False), daemon=True, name="foo").start()

# Wait for Flask to start
time.sleep(1)

# Continuous request loop
while True:
    requests.get("http://127.0.0.1:8080/foo", timeout=30)
```
### Summary of Fixes

| **Issue**                          | **Fix**                                                                 |
|------------------------------------|-------------------------------------------------------------------------|
| **Multi-threading instability**    | Forced single-threaded OpenVINO execution + Flask running with `threaded=False` |
| **Floating-point drift / denormals**| Enabled `OPENVINO_DENORMALS_ZERO=1`                                     |
| **Inference instability over time** | Re-loading the model on each inference call                             |
| **Precision variability**          | Explicitly using `INFERENCE_PRECISION_HINT=f32`                         |

Let me know if the issue persists after these changes!

**Comment by Dobiasd** [comment]
Hi, @rkazants, and thanks for looking into it.

Yes, without Docker, the effect no longer happens on my machines either.

I just re-ran the test with your suggestion included (set `share_weights=False` in the `ov.convert_model` call), and the effect no longer happened (at least not up to the number of iterations I let the process run). :tada:

I'll implement this workaround for our PROD application. 👍

**Comment by Dobiasd** [comment]
Hi, @Bheema-Shanker-Neyigapula,

thanks for the (LLM-generated? :wink:) suggestions.

> 1. Forcing Single-Threaded Execution

This is a bit too slow.

> 2. Handling Denormal Values

Ok.

> 3. Ensuring a Fresh Model Instance on Each Inference Call

This is way too slow.

> 4. Using Explicit NumPy Dtype (`float32`)

Ok.

[Using only suggestions 2 and 4 (and not 1 and 3)](https://gist.github.com/Dobiasd/4a775f5dbef7bc6b5eb0b56299f5975a), the problem still exists.

For now, I'll go with the `share_weights=False` workaround.

**Comment by Dobiasd** [comment]
Hi, @Bheema-Shanker-Neyigapula,

thanks for the (LLM-generated? :wink:) suggestions.

> 1. Forcing Single-Threaded Execution

This is a bit too slow.

> 2. Handling Denormal Values

Ok.

> 3. Ensuring a Fresh Model Instance on Each Inference Call

This is way too slow.

> 4. Using Explicit NumPy Dtype (`float32`)

Ok.

[Using only suggestions 2 and 4 (and not 1 and 3)](https://gist.github.com/Dobiasd/4a775f5dbef7bc6b5eb0b56299f5975a), the problem still exists.

For now, I'll go with the `share_weights=False` workaround.

**Comment by Bheema-Shanker-Neyigapula** [comment]
Hi @Dobiasd ,  

No LLM was used—I was just exploring different options.  

Got it on the performance concerns. If suggestions 2 and 4 alone don’t resolve the issue, the `share_weights=False` workaround makes sense for now. Let me know if you want to dig deeper into other optimizations!

**Comment by Dobiasd** [comment]
Ok, thanks @Bheema-Shanker-Neyigapula! (And sorry for the false suspicion.)

I just deployed the modified version of our PROD application and will monitor the stability there too. 👍

**Comment by Dobiasd** [comment]
@rkazants Oh, why did you close this issue? There is a workaround, but using `share_weights=True` (the default) is still broken, or am I missing something?

**Comment by Dobiasd** [comment]
Ah! 💡 Got it. Thanks a lot for the explanation. 👍

Would it make sense to change the default value of `share_weights` from `True` to `False` to prevent other people from making the same mistake that I made?

**Comment by rkazants** [comment]
@Dobiasd, `share_weights=True ` is a important feature to avoid doubled-size model in memory (original + OV instances) that is critical for modern huge models. We want to provide users capability to convert such models on standard client devices.
We know it is tricky case that is why we have documented these cautions.

Best regards,
Roman

**Comment by mlukasze** [comment]
btw, in upcoming (soon) release we will change default value of this attribute to `true`
in case you will prefer `false`, for any reason, please take advice to set up specificaly, to avoid unpredicted behaviors after OV upgrade

**Comment by rkazants** [comment]
it has been already set to `True` several releases ago:)

![Image](https://github.com/user-attachments/assets/7dd1a766-5c62-485e-b2cc-80154244d882)

Best regards,
Roman

**Comment by mlukasze** [comment]
oh, here, you are right 👍 
mixed up with [changes for Tensors](https://github.com/openvinotoolkit/openvino/pull/27905)
